### ISSUE TITLE
PostCSS 7 compat build dependency requiring PostCSS 8

### ISSUE THREAD

**Issue by jorditarrida** [tailwindlabs/tailwindcss#3101]
### Describe the problem:

It looks like `@fullhuman/postcss-purgecss`, which is a Tailwind dependency (^3.0.0), was updated to 3.1.0 recently. This version has `purgecss@^3.1.0` as a dependency, which in turn has `postcss@^8.2.1` as a dependency. So whenever I try to install the Tailwind PostCSS 7 compatibility build on a new project, I get `@fullhuman/postcss-purgecss@3.1.0` installed, and the purge process returns an error which I have seen when using wrong versions of PostCSS and its plugins: `[object Object] is not a PostCSS plugin`.

Maybe locking the version to `@fullhuman/postcss-purgecss@3.0.0` would solve the issue?

I am currently teaching Tailwind at the [Open University of Catalonia (UOC)](https://www.uoc.edu/) and some students are currently facing this issue. I tested the tools thoroughly when Tailwind 2 was released to make the switch for the students (which I was aware was a risky move), and it appeared to work, but it seems it broke a few days ago because of this. So could you please check if this is an issue and if there would be a quick way to fix it? (like recommending locking to `@fullhuman/postcss-purgecss@3.0.0`)

I also noted on the documentation the installation for the compatibility build was changed from `npm install tailwindcss@compat` to `npm install tailwindcss@npm:@tailwindcss/postcss7-compat @tailwindcss/postcss7-compat`. Is this related to this issue in any way?

### Link to a minimal reproduction:

Reproduction steps would be installing the compatibility build on any project and checking that `@fullhuman/postcss-purgecss@3.1.0` was installed, which has `purgecss@^3.1.0` as a dependency, which in turn has `postcss@^8.2.1` as a dependency.

I created [this repo](https://github.com/jorditarrida/uoc-boilerplate-tailwind-test) a week ago. You can see on the [lockfile](https://github.com/jorditarrida/uoc-boilerplate-tailwind-test/blob/master/package-lock.json) the version it got installed was `@fullhuman/postcss-purgecss@3.0.0`. But following the same steps today (cloning [this repo](https://github.com/uoc-advanced-html-css/uoc-boilerplate) and installing Tailwind on top of it) would today install `@fullhuman/postcss-purgecss@3.1.0` which I believe causes this issue.

**Comment by adamwathan** [comment]
I can't seem to reproduce this — here is a link to a dead simple repository using the compatibility build along with PurgeCSS 3.1 and running a production build (`npm run build`) builds successfully without any errors:

https://github.com/adamwathan/tw-postcss7-purge-test

Can you create a similar repository that I can use to reproduce? Ideally something I can literally just clone and run `npm install` then some build command to see the error, rather than a list of instructions I need to follow to create the project myself.

**Comment by adamwathan** [comment]
Looks like adding the `--legacy-peer-deps` flag solves the issue — guessing you are running npm 7.

```
npm install --legacy-peer-deps tailwindcss@npm:@tailwindcss/postcss7-compat @tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9
```

I will raise on issue on PurgeCSS about removing PostCSS 8 as an explicit peer dependency if they plan to continue supporting PostCSS 7 at the same time.

**Comment by adamwathan** [comment]
I still am not sure why I can't recreate this in a simple PostCSS 7 project, which is unfortunate because it prevents me from providing a reproduction to @Ffloriel on the PurgeCSS repo. Generally though I think the fix is to remove `postcss` as an explicit peer-dependency over there until support for PostCSS 7 is dropped.

**Comment by jorditarrida** [comment]
Hi @adamwathan,
As always thanks for your prompt reply and help.
I just created the following repo for you to check: https://github.com/jorditarrida/tailwind-compat-purgecss-test
(As I stated, what I'm doing is cloning the base Parcel v1 based project used by the students, and installing Tailwind on top of it).
As you will see, `@fullhuman/postcss-purgecss@3.1.0` was installed (compared to [this other repo](https://github.com/jorditarrida/uoc-boilerplate-tailwind-test) I created just last week which got 3.0.0 following the same steps - and works), and when you run `npm run build` it gives the `[object Object] is not a PostCSS plugin` error.
However, if you comment out the `purge` option in `tailwind.config.js`, it builds correctly.
I am recommending to my students using the Node.js LTS version (14.15.1), so they are still using npm 6(.14.8).
Thanks again for checking this out :)

**Comment by Ffloriel** [comment]
Yes, it makes sense to remove postCSS 8 as a peer dependency since version 7 is still supported.
Thanks for the repo @jorditarrida , it looks like the issue is with PurgeCSS. I'll look into it.

**Comment by arlyon** [comment]
Just to add to this, while it doesn't break anything, tailwindcss depends on postcss-js which as of 3.0.2 also uses postcss 8.

**Comment by mattrothenberg** [comment]
I can replicate this in a `parcel` project as well. My `yarn.lock` looks like the following after install TW's compatibility build (note the dependency on PostCSS 8 in the latter definition)

```
purgecss@^2.3.0:
  version "2.3.0"
  resolved "https://registry.yarnpkg.com/purgecss/-/purgecss-2.3.0.tgz#5327587abf5795e6541517af8b190a6fb5488bb3"
  integrity sha512-BE5CROfVGsx2XIhxGuZAT7rTH9lLeQx/6M0P7DTXQH4IUc3BBzs9JUzt4yzGf3JrH9enkeq6YJBe9CTtkm1WmQ==
  dependencies:
    commander "^5.0.0"
    glob "^7.0.0"
    postcss "7.0.32"
    postcss-selector-parser "^6.0.2"

purgecss@^3.1.3:
  version "3.1.3"
  resolved "https://registry.yarnpkg.com/purgecss/-/purgecss-3.1.3.tgz#26987ec09d12eeadc318e22f6e5a9eb0be094f41"
  integrity sha512-hRSLN9mguJ2lzlIQtW4qmPS2kh6oMnA9RxdIYK8sz18QYqd6ePp4GNDl18oWHA1f2v2NEQIh51CO8s/E3YGckQ==
  dependencies:
    commander "^6.0.0"
    glob "^7.0.0"
    postcss "^8.2.1"
    postcss-selector-parser "^6.0.2"
```

**Comment by MarcoMacedoP** [comment]
It's there any update on this? I'm facing with the same issue

**Comment by joejordan** [comment]
Using Node v14.15.3, I attempted to install Tailwind 2.0.2 with the `--legacy-peer-deps` option as @adamwathan recommended; that did not resolve the issue, as including any purge options in tailwind.config caused the PostCSS 8 requirement error upon build.

I also attempted @mattrothenberg's suggestion to add a postcss resolution; that also did not resolve the issue.

For the time being I am purging CSS externally and setting `purge: false` in the tailwind.config so that I can continue using Tailwind v2.

**Comment by kelvinz** [comment]
@joejordan How are you purging externally? I'm also having this issue and am unable to build my project. =(

**Comment by jorisnoo** [comment]
@joejordan I was able to build the project (based on Parcel) by explicitly requiring `"@fullhuman/postcss-purgecss": "3.0"` as a dependency.

**Comment by illepic** [comment]
For anyone looking for the quick fix: `npm i -D @fullhuman/postcss-purgecss@3.0` and then `npm run build`.

**Comment by RichardJECooke** [comment]
I don't know if this is exactly the same problem, but this worked for me: https://github.com/tailwindlabs/tailwindcss/discussions/2854

**Comment by Tesha0200** [comment]
Autoprefixer uses the new PostCSS 8 API since version 10.  I am not sure about this but can you try installing posts css as a dependency?
